compiler/rustc_data_structures/src/jobserver.rs

```diff
@@ -1,40 +1,67 @@
 pub use jobserver_crate::Client;
-use std::sync::LazyLock;
-
-// We can only call `from_env` once per process
-
-// Note that this is unsafe because it may misinterpret file descriptors
-// on Unix as jobserver file descriptors. We hopefully execute this near
-// the beginning of the process though to ensure we don't get false
-// positives, or in other words we try to execute this before we open
-// any file descriptors ourselves.
-//
-// Pick a "reasonable maximum" if we don't otherwise have
-// a jobserver in our environment, capping out at 32 so we
-// don't take everything down by hogging the process run queue.
-// The fixed number is used to have deterministic compilation
-// across machines.
-//
-// Also note that we stick this in a global because there could be
-// multiple rustc instances in this process, and the jobserver is
-// per-process.
-static GLOBAL_CLIENT: LazyLock<Client> = LazyLock::new(|| unsafe {
-    Client::from_env().unwrap_or_else(|| {
+
+use jobserver_crate::FromEnvErrorKind;
+
+use std::sync::{LazyLock, OnceLock};
+
+// We can only call `from_env_ext` once per process
+
+// We stick this in a global because there could be multiple rustc instances
+// in this process, and the jobserver is per-process.
+static GLOBAL_CLIENT: LazyLock<Result<Client, String>> = LazyLock::new(|| {
+    // Note that this is unsafe because it may misinterpret file descriptors
+    // on Unix as jobserver file descriptors. We hopefully execute this near
+    // the beginning of the process though to ensure we don't get false
+    // positives, or in other words we try to execute this before we open
+    // any file descriptors ourselves.
+    let jobserver_crate::FromEnv { client, var } = unsafe { Client::from_env_ext(true) };
+
+    let error = match client {
+        Ok(client) => return Ok(client),
+        Err(e) => e,
+    };
+
+    // Create jobserver if it wasn't exist.
+    if matches!(
+        error.kind(),
+        FromEnvErrorKind::NoEnvVar | FromEnvErrorKind::NoJobserver | FromEnvErrorKind::Unsupported
+    ) {
+        // Pick a "reasonable maximum" capping out at 32
+        // so we don't take everything down by hogging the process run queue.
+        // The fixed number is used to have deterministic compilation across machines.
         let client = Client::new(32).expect("failed to create jobserver");
+
         // Acquire a token for the main thread which we can release later
         client.acquire_raw().ok();
-        client
-    })
+
+        return Ok(client);
+    }
+
+    // Environment specifies jobserver, but it looks incorrect.
+    // Safety: `error.kind()` should be `NoEnvVar` if `var == None`.
+    let (name, value) = var.unwrap();
+    Err(format!("failed to connect jobserver from `{name}={:?}`: {error}", value.to_str()))
```

Comment:
```suggestion
    Err(format!("failed to connect to jobserver from environment variable `{name}={:?}`: {error}", value))
```

Reply:
Could you also add an `early_note` to this error saying "the build environment is likely misconfigured"?

Reply:
Do you mean changing error message here or calling `early_note` before `early_warning`?

Reply:
I mean attaching a note to the warning.